modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/ExecutorImpl.scala

```diff
@@ -3,66 +3,99 @@ package edu.gemini.seqexec.server
 import edu.gemini.seqexec.odb.SeqExecService
 import edu.gemini.spModel.core.Peer
 
-import scalaz._
+import scalaz.{\/, _}
 import Scalaz._
-
 import edu.gemini.pot.sp.SPObservationID
 import edu.gemini.seqexec.server.Executor._
 import edu.gemini.seqexec.server.SeqexecFailure._
 import edu.gemini.spModel.config2.ConfigSequence
 
 import scalaz.concurrent.Task
+import scalaz.stream.Process
+import scalaz.stream.async
 
 /** An executor that maintains state in a pair of `TaskRef`s. */
 class ExecutorImpl private (cancelRef: TaskRef[Set[SPObservationID]], stateRef: TaskRef[Map[SPObservationID, Executor.ExecState]]) {
+  type ExecutionFailures = NonEmptyList[SeqexecFailure] \/ Unit
+  type ExecutionResult = (ExecState, ExecutionFailures)
 
   private var loc = new Peer("localhost", 8443, null)
 
+  // Queue to keep events generated when executing a sequence
+  private val eventsQueue = async.topic[SeqexecEvent]()
+
   def host(): Peer = loc
   def host(l: Peer): Unit = { loc = l }
 
-  private def recordState(id: SPObservationID)(s: ExecState): Task[Unit] =
-    stateRef.modify(_ + (id -> s))
+  def stepEvent(id: SPObservationID, s: ExecState): SeqexecEvent = s.completed.reverse match {
+    case Nil        => SequenceStartEvent(id.toString)
+    case Ok(r) :: _ => StepExecutedEvent(id.toString, s.completed.size, s.remaining.size, r.observeResult.dataId)
+    case l          => NullEvent //Fallback
+  }
+
+  private def recordState(id: SPObservationID)(s: ExecState): Task[ExecState] =
+    // This marks the current state for observation id
+    eventsQueue.publishOne(stepEvent(id, s)) >> stateRef.modify(_ + (id -> s)) >> Task.delay(s)
```

Comment:
This seems Good. Stylistically you should prefer `*>` over `>>` because it is less constrained (it only requires `Apply` but `>>` requires `Bind` ... there is really never any reason to use `>>`).